capstone/src/main/java/com/google/sps/servlets/GroupPostDataServlet.java

```diff
@@ -0,0 +1,93 @@
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.SortDirection;
+import com.google.appengine.api.datastore.Key;
+import com.google.appengine.api.datastore.KeyFactory;
+import com.google.appengine.api.users.UserService;
+import com.google.appengine.api.users.UserServiceFactory;
+import com.google.appengine.api.images.ImagesService;
+import com.google.appengine.api.images.ImagesServiceFactory;
+import com.google.appengine.api.images.ServingUrlOptions;
+import java.io.PrintWriter;
+import java.net.MalformedURLException;
+import java.net.URL;
+import com.google.sps.Objects.User;
+import com.google.sps.Objects.Post;
+import com.google.sps.Objects.Comment;
+
+@WebServlet("/group-post")
+public class GroupPostDataServlet extends HttpServlet {
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+
+    Query query = new Query("Post").addSort("timestamp", SortDirection.DESCENDING);
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    PreparedQuery results = datastore.prepare(query);
+
+    List<Post> posts = new ArrayList<>();
+    for (Entity entity : results.asIterable()) {
+      posts.add(getPostEntity(entity));
+    }
+
+    // Convert to json
+    response.setContentType("application/json;");
+    response.getWriter().println(new Gson().toJson(posts));
+  }
+
+  public Post getPostEntity(Entity entity) {
+    long timestamp = (long) entity.getProperty("timestamp");
+    String authorId = (String) entity.getProperty("authorId");
+    String postText = (String) entity.getProperty("postText");
+    String challengeName = (String) entity.getProperty("challengeName");
+    String img = (String) entity.getProperty("img");
+    HashSet<String> likes = (HashSet<String>) entity.getProperty("likes");
+    ArrayList<Comment> comments = (ArrayList<Comment>) entity.getProperty("comments");
+    Post userPost = new Post(authorId, postText, comments, challengeName, timestamp, img, likes);
+    return userPost;
+  }
+
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+
+    // Receives submitted post 
+    long timestamp = System.currentTimeMillis();
+    String authorId = "Jane Doe";
+    String postText = request.getParameter("post-input");
+    String challengeName = "Challenge Name";
+    String img = "";
+    HashSet<String> likes = new HashSet<>();
+    ArrayList<Comment> comments = new ArrayList<>();
+
+    // Creates entity with submitted data and add to database
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    datastore.put(createPostEntity(timestamp, authorId, postText, challengeName, img, likes, comments));
+
+    // Redirect back to the HTML page.
+    response.sendRedirect("/group.html");
+  }
```

Comment:
I just wanted to ask, based on comments made by Lucy and Himani, does datastore support casting to a HashSet (line 57) and putting a HashSet (line 76/89)? This builds and doesn't produce any errors?

Reply:
I know Datastore supports putting a HashSet. I'm unsure about casting, I think that was the main issue when Lucy tried it.